### ISSUE TITLE
Questioning weakref use in parent

### ISSUE THREAD

**Issue by pmcharrison** [music-computing/amads#45]
We have this:

```py
    @parent.setter
    def parent(self, p):
        self._parent = weakref.ref(p)
```

As I understand it, saving `parent` as a weakref means that Python will garbage-collect the parent if the parent is not saved separately as a variable. So, in this code:

```py
note = read_midi("...").notes[0]
```

the `note.parent` attribute will potentially be unavailable because it the parent might have been garbage collected. Is this what we want? Maybe @rbdannenberg can comment?

**Comment by rbdannenberg** [comment]
I was under the impression the Python does not reclaim structures with cycles due to reference counting. But in learning about weak references, it seems there is some kind of tracing garbage collection in addition to reference counting, but it's unclear if relying on that is a reasonable thing, given all the warnings about cycles and reference counting. If cycles can actually be reclaimed (and there's no problem with it -- it might be expected to occur a lot), then I think we should make parent an ordinary reference. Good question! I would strongly recommend against keeping a reference to a note if there's no reference to the parents. If the note has some use outside of the parents, it should probably be copied into another score structure, freeing the original note.

**Comment by pmcharrison** [comment]
Yes, I do think Python has a cycle detector, see the last paragraph of [this page](https://docs.python.it/html/ext/refcounts.html). Though it is possible to configure a Python installation with the cycle detector I think this is a perverse use case... So probably it is best just to not worry and rely on the built-in cycle detector?

**Comment by rbdannenberg** [comment]
I think you mean it is possible to configure Python _without_ the cycle detector, but this is a perverse case. I agree. So I now think it's best to eliminate the weak reference to parent and just allow the cycles it will create. If you create scores without cycle detection, they will not be freed automatically, so you'll leak memory, but even then, it's unlikely anyone will notice given the size of scores. Now I'm curious about how much memory Python objects take -- it would be interesting to know the approximate size of a score with 1000 notes just for back-of-the-envelope calculations.

**Comment by pmcharrison** [comment]
Sorry, you're right, I meant without! I agree, it would be useful to know how much memory the objects take. Will see if I can find that out sometime.

**Comment by diwangislucky** [comment]
I was under the impression that, while cycles are able to be handled by the garbage collection, it takes longer for the garbage collector to detect and free cyclic references. This is why the discussion I've seen on this topic seems to allow weak references as an acceptable compromise to break cyclic references. However, given the size of the scores, not sure if it's worth it worrying over a longer memory lifecycle?

**Comment by rbdannenberg** [comment]
That sounds possible -- certainly the "eager" freeing by ref counts would happen sooner than waiting for a tracing GC, and maybe since garbage doesn't build fast due to ref counts, whatever Python does for cycle removal might not need to run very often. I think we should go with strong references and see what happens: Not freeing scores at all would be OK for almost anything but very large batch processing of datasets.

**Comment by pmcharrison** [comment]
Sounds good to me!
